test/session-pool.js

```diff
@@ -16,2022 +16,1575 @@
-          numTimesReleased++;
-          assert.strictEqual(session, SESSION);
+      it('should handle when invalid session is passed', function() {
+        let keptAlive = false;
+        let deleted = false;
+        writeSession.keepAlive = function() {
+          keptAlive = false;
+          return Promise.reject();
+        };
+        writeSession.delete = function() {
+          deleted = true;
+          return Promise.resolve();
         };
 
-        var requestStream = sessionPool.requestStream(CONFIG);
-
-        requestStream.emit('reading');
-
-        setImmediate(function() {
-          requestStream.abort();
-          assert.strictEqual(numTimesReleased, 1);
-
-          requestStream.abort();
-          assert.strictEqual(numTimesReleased, 1);
-
-          done();
+        sessionPool.sendKeepAlive_().then(function() {
+          assert.strictEqual(keptAlive, false);
+          assert.strictEqual(deleted, false);
         });
       });
+    });
+  });
 
-      it('should cancel the request stream', function(done) {
-        REQUEST_STREAM.cancel = done;
-
-        var requestStream = sessionPool.requestStream(CONFIG);
-
-        requestStream.emit('reading');
+  describe('pingSession', function() {
+    let sessionPool;
+    beforeEach(function() {
+      sessionPool = new SessionPool(DATABASE, {
+        maxReads: 10,
+        minReads: 10,
+        minWrites: 5,
+        keepAlive: 0.001,
+      });
+    });
 
-        setImmediate(function() {
-          requestStream.abort();
+    it('should ping min read and write session', function() {
+      let readKeepAlive = 0;
+      let writeKeepAlive = 0;
+      sessionPool.createTransaction_ = () => Promise.resolve();
+      sessionPool.sendKeepAlive_ = s => {
+        if (s.type === 'readonly') {
+          readKeepAlive++;
+        } else {
+          writeKeepAlive++;
+        }
+        sessionPool.release(s);
+        return Promise.resolve();
+      };
+      sessionPool.session_ = function() {
+        return new Object({
+          create: () => Promise.resolve(),
+          delete: () => Promise.resolve(),
         });
+      };
+      sessionPool.open();
+      // Clear timeout open() call sets
+      // sessionPool.pingSessions_();
+      return new Promise(resolve => {
+        setTimeout(() => {
+          clearTimeout(sessionPool.pingTimeoutHandle);
+          assert.strictEqual(readKeepAlive, 20);
+          assert.strictEqual(writeKeepAlive, 10);
+          return resolve();
+        }, 150);
       });
     });
-  });
 
-  describe('size', function() {
-    it('should return the total number of sessions', function() {
-      sessionPool.reads_ = [{}, {}];
-      sessionPool.borrowed_ = [{}, {}, {}];
-
-      assert.strictEqual(sessionPool.size(), 5);
+    afterEach(function() {
+      return sessionPool.close();
     });
   });
 
-  describe('acquireSession_', function() {
+  describe('should convert session types', function() {
```

Comment:
Can we move these tests for the corresponding methods? Having them in a top level describe block makes them harder to navigate.

Reply:
Fixed it